File: iosApp/iosApp/Pages/Map/ChildStopLayerGenerator.swift
//
//  ChildStopLayerGenerator.swift
//  iosApp
//
//  Created by Horn, Melody on 2024-06-24.
//  Copyright © 2024 MBTA. All rights reserved.
//

import MapboxMaps
import shared

class ChildStopLayerGenerator {
    let childStopLayer: MapboxMaps.SymbolLayer = createChildStopLayer()

    static let childStopLayerId = "child-stop-layer"

    static let annotationTextZoomThreshold = 19.0

    static func createChildStopLayer() -> MapboxMaps.SymbolLayer {
        var layer = MapboxMaps.SymbolLayer(
            id: Self.childStopLayerId,
            source: ChildStopFeaturesBuilder.shared.childStopSourceId
        )
        layer.iconImage = .expression(Exp(.match) {
            Exp(.get) { ChildStopFeaturesBuilder.shared.propLocationTypeKey }
            String(describing: LocationType.entranceExit)
            ChildStopIcons.entranceIcon
            [String(describing: LocationType.boardingArea), String(describing: LocationType.stop)]
            ChildStopIcons.platformIcon
            ""
        })
        layer.textField = .expression(Exp(.step) {
            Exp(.zoom)
            ""
            Self.annotationTextZoomThreshold
            Exp(.get) { ChildStopFeaturesBuilder.shared.propNameKey }
        })

        layer.textColor = .constant(.init(.gray))
        layer.textFont = .constant(["Inter Italic"])
        layer.textHaloColor = .constant(.init(.fill3))
        layer.textHaloWidth = .constant(2.0)
        layer.textSize = .constant(12)
        layer.textVariableAnchor = .constant([.left, .right, .bottom, .top])
        layer.textJustify = .constant(.auto)
        layer.textOptional = .constant(true)
        layer.textRadialOffset = .constant(1)
        layer.iconAllowOverlap = .constant(true)
        layer.textAllowOverlap = .constant(false)
        layer.symbolSortKey = .expression(Exp(.get) { ChildStopFeaturesBuilder.shared.propSortOrderKey })
        layer.minZoom = MapDefaults.shared.closeZoomThreshold
        return layer
    }
}

File: shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/map/ChildStopLayerGenerator.kt
package com.mbta.tid.mbta_app.map

import com.mbta.tid.mbta_app.map.style.Exp
import com.mbta.tid.mbta_app.map.style.SymbolLayer
import com.mbta.tid.mbta_app.map.style.TextAnchor
import com.mbta.tid.mbta_app.map.style.TextJustify
import com.mbta.tid.mbta_app.map.style.downcastToColor
import com.mbta.tid.mbta_app.map.style.downcastToResolvedImage
import com.mbta.tid.mbta_app.model.LocationType

internal object ChildStopLayerGenerator {
    val childStopLayerId = "child-stop-layer"

    val annotationTextZoomThreshold = 19.0

    fun createChildStopLayer(colorPalette: ColorPalette): SymbolLayer {
        var layer =
            SymbolLayer(id = childStopLayerId, source = ChildStopFeaturesBuilder.childStopSourceId)
        layer.iconImage =
            Exp.match(
                    Exp.get(ChildStopFeaturesBuilder.propLocationTypeKey),
                    Exp(LocationType.ENTRANCE_EXIT.name) to Exp(ChildStopIcons.entranceIcon),
                    Exp.Bare.arrayOf(LocationType.BOARDING_AREA.name, LocationType.STOP.name) to
                        Exp(ChildStopIcons.platformIcon),
                    fallback = Exp(""),
                )
                .downcastToResolvedImage()
        layer.textField =
            Exp.step(
                Exp.zoom(),
                Exp(""),
                Exp(annotationTextZoomThreshold) to Exp.get(ChildStopFeaturesBuilder.propNameKey),
            )

        // TODO actually pick a color
        layer.textColor = Exp("#888888").downcastToColor()
        layer.textFont = listOf("Inter Italic")
        layer.textHaloColor = Exp(colorPalette.fill3).downcastToColor()
        layer.textHaloWidth = 2.0
        layer.textSize = 12.0
        layer.textVariableAnchor =
            listOf(TextAnchor.LEFT, TextAnchor.RIGHT, TextAnchor.BOTTOM, TextAnchor.TOP)
        layer.textJustify = TextJustify.AUTO
        layer.textOptional = true
        layer.textRadialOffset = 1.0
        layer.iconAllowOverlap = true
        layer.textAllowOverlap = false
        layer.symbolSortKey = Exp.get(ChildStopFeaturesBuilder.propSortOrderKey)
        layer.minZoom = MapDefaults.closeZoomThreshold
        return layer
    }
}
